### DISCUSSION TITLE
Brainstorming ideas for community contributions

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [Copilot-Language/copilot#557] [General]
I'd like to open a thread to locate ideas of issues that would help the Copilot project that would be good candidates for new contributors, including some with no prior Haskell experience.

I see Copilot as a community project, and I'd like to increase the number of participants. I think that will be a win-win for the Haskell Community and for the project itself.

Please add below any ideas of improvements that would be great for external contributors.

**Note: Please don't just start working on these ideas and start sending PRs. Our project tracks issues and PRs in a special way. Comment on the idea if you think you'd like to contribute that.**

**Comment by ivanperez-keera** [comment]
The banner for the build status doesn't show up right on my machine. Github has cached the result but it's not providing the resulting image, so we see "Build Status" instead:

![image](https://github.com/user-attachments/assets/b63e9055-1d11-4ba5-b002-9c84b7fa2297)

**Reply by fdedden** [reply]
I can confirm this, doesn't work for me either.

**Reply by RyanGlScott** [reply]
I think the [modern URL](https://docs.travis-ci.com/user/status-images/) for the badge is https://api.travis-ci.com/Copilot-Language/copilot.svg?branch=master :

![Build status](https://api.travis-ci.com/Copilot-Language/copilot.svg?branch=master)

**Reply by ivanperez-keera** [reply]
This would be a good candidate for a new-comer who's never submitted patches to the project.

**Reply by ivanperez-keera** [reply]
I'd love to add this to the next release, Copilot 4.3.

**Reply by kaBeech** [reply]
I've been leaving this for a newcomer to the project, but I'd be happy to do it if you like!

**Comment by ivanperez-keera** [comment]
We could remove the fields in `UExpr` similarly to how we are doing it for `UType` in https://github.com/Copilot-Language/copilot/issues/484.

**Reply by kaBeech** [reply]
I'm interested in helping with this!

**Reply by ivanperez-keera** [reply]
Hi @kaBeech . I've created an issue with details: https://github.com/Copilot-Language/copilot/issues/565.

I'm hoping we can complete this for Copilot 4.2, which we will release on Jan 7, 2025.

The agreement you'll have to sign before the PR can be accepted will be very similar to https://github.com/NASA-SW-VnV/ikos/tree/master/doc/contribute (don't sign that one: that's for a different project, it's just to illustrate). One of those is if you are contributing as your work for a company ("Corporate"), the other is if you are contributing in your individual capacity ("Individual"). The latter is the most common scenario. You'll only need to sign one or the other.

Let me know if this looks good and if you think you can complete that before Dec 21st, 2025. If so, I can assign the issue to you and I can guide you through the process and through any changes needed to accept the PR.

Thanks!

**Reply by ivanperez-keera** [reply]
To clarify: you'll only need to sign the agreement once. You won't have to do that with every PR, just the first.

**Reply by kaBeech** [reply]
Thanks @ivanperez-keera, this all looks good! I am confident I can complete that before Dec 21st, 2025. I am ready to be assigned!

**Reply by ivanperez-keera** [reply]
@kaBeech -- thanks for volunteering to help with this issue.

I'm going to try to list in detail what I think is needed (I may have forgotten something). Let me know if something's unclear.

- Create a branch `develop-deprecate-UExpr-fields` forking off of our current `master`.
- Commit 1: `copilot-c99`: Remove any uses of `uExprType` and `uExprExpr`. Use positional arguments instead. For example, in `copilot-c99/src/Copilot/Compile/C99/CodeGen.hs`, the code:
```haskell
            declare :: UExpr -> C.Ident -> C.Decln                               
            declare (UExpr { uExprType = ty }) tmpVar = 
```
could become:
```haskell
            declare :: UExpr -> C.Ident -> C.Decln                               
            declare (UExpr ty _) tmpVar = 
```

- Commit 2: `copilot-prettyprinter`: Remove any uses of `uExprType` and `uExprExpr`.
- Commit 3: `copilot-theorem`: Remove any uses of `uExprType` and `uExprExpr`.
- Commit 4: `copilot-core`: Deprecate `uExprType` and `uExprExpr`.
- Commit 5: `copilot-c99`: Document changes in CHANGELOG. If there isn't a block of entries for the next release, you can create one at the top with the current date. Otherwise, you can add your entry at the top of the latest block.
- Commit 6: `copilot-prettyprinter`: Document changes in CHANGELOG. Ditto.
- Commit 7: `copilot-theorem`: Document changes in CHANGELOG. Ditto.
- Commit 8: `copilot-core`: Document changes in CHANGELOG. Ditto.

We are quite thorough and systematic in the way we document commits. You can take a look at the commit messages in #556 and #483 to see how we do things. Feel free to email me at `ivan.perezdominguez@nasa.gov` if you have any questions; I'm also happy to jump on a video call.

When you start working on this, please let me know.

When you are ready for me to review your PR, you can write a comment in issue #565 like this one:

https://github.com/Copilot-Language/copilot/issues/564#issuecomment-2473832227

This lets me know that you are ready for me to take a look and perform any testing needed on our side.

If you are not entirely sure but you want me to take a look at your current draft, I'm happy to do that too.

Thanks again for your help!

**Reply by ivanperez-keera** [reply]
Btw, in case it helps, I think these commits are 1-3 lines each. It's not a huge patch overall. It's just the number of commits that may make it look bigger than it actually is.

**Reply by kaBeech** [reply]
Perfect, that process is about what I expected from looking at #484. I'll get started this afternoon (after I finish another project) and I'll ping you when I start actually working on it. I'll email you as well so you have my private address (I get a lot of spam in my public one).

Thank you for taking such care in getting me set up to work on this! It's a pleasure to be working on this project. I'm curious about Copilot's utility in other fields too, such as audio processing. That's likely a topic more appropriate for email, but needless to say I'm excited to contribute!

**Reply by kaBeech** [reply]
> 
> When you start working on this, please let me know.
> 

Hey @ivanperez-keera, I'm starting work on this!

**Reply by kaBeech** [reply]
Oh, shoot, looking at Travis CI for my PR #567 it looks like I mis-indented my deprecation warning. What is the best way to fix this? Can I rebase my commit since it hasn't been merged yet, or should I make a new commit that fixes it?

**Reply by ivanperez-keera** [reply]
No worries. It happens.

Since I haven't started reviewing it yet, I propose that you remove this comment:
https://github.com/Copilot-Language/copilot/issues/565#issuecomment-2477856008
and keep working on your PR until it's ready. When you have the PR building on Travis and you want me to take a look, you can post the comment again.

Yes, rebasing is the right way to go.

FYI, I also suspect this may be a syntax error (I'd expect you to need parentheses around the `UEXpr` and its arguments):

https://github.com/Copilot-Language/copilot/blob/99fc9757f88f6dcd500747d4f64c66a0b22fdb05/copilot-prettyprinter/src/Copilot/PrettyPrint.hs#L46

**Reply by ivanperez-keera** [reply]
In case it helps: Since the CI jobs take about 20 minutes to complete, normally I just build everything on my machine first and then review the PR with `gitk` before sending it. That way, if something's broken, I don't have to wait for the CI job to crash to find out.

**Reply by kaBeech** [reply]
Thank you! Will do.

**Reply by kaBeech** [reply]
Thank you again for your patience - those were some very silly mistakes made in haste. I have improved my workflow to reduce such oversights in the future. Good point about reviewing my PR with gitk as well - for some reason it had not occurred to me that I could do that locally before pushing.

It looks good to me now (fingers crossed) and it both builds locally and on Travis - I've re-posted my comment.

**Reply by ivanperez-keera** [reply]
No problem. Thanks for sending the improved version so quickly!

At first glance it looks good. I'll do a closer review soon.

**Reply by ivanperez-keera** [reply]
Massive thanks to @kaBeech for contributing a fix for this issue!

**Reply by kaBeech** [reply]
Thank you for guiding me through the process @ivanperez-keera! It's been an absolute pleasure and I look forward to contributing more.

**Comment by ivanperez-keera** [comment]
Copilot is now compatible with GHC 9.10. The README could be updated to reflect that.

**Reply by kaBeech** [reply]
I can help with this!

Edit: I got excited and volunteered for most of these. If anyone else is excited to contribute and wants to jump in on one or two of the ones I commented on, please feel free - I don't mean to hog them all!

**Reply by ivanperez-keera** [reply]
Haha. No problem. I've commented on one to see if you want to contribute that one first :)

**Reply by ivanperez-keera** [reply]
I'd love to add this to the next release, Copilot 4.3.

**Comment by ivanperez-keera** [comment]
The installation instructions are a bit odd. They tell the user to call the interpreter with an import of a Copilot module, but then the output is not very easy to understand.

**Reply by ivanperez-keera** [reply]
A separate thread was opened before to track how to best do this: https://github.com/Copilot-Language/copilot/discussions/543.

**Reply by kaBeech** [reply]
I'm interested in helping with this.

**Comment by ivanperez-keera** [comment]
We'd like Copilot to have homebrew packages so that people can just install it on Mac with `brew install copilot`.

**Reply by ivanperez-keera** [reply]
A separate thread was opened before to track how to best do this: https://github.com/Copilot-Language/copilot/discussions/537.

**Comment by ivanperez-keera** [comment]
This module https://hackage.haskell.org/package/copilot-core-4.0/docs/Copilot-Core-Type.html could use haddock headers to structure the documentation page into Types, Arrays, Structs, etc.

**Reply by kaBeech** [reply]
I'm interested in helping with this.

It seems like most of the information would be under the Types and Structs headers, with a few exceptions. It seems like `typeSize` and `typeLength` would be under the Array header. Is this correct?

**Reply by ivanperez-keera** [reply]
I'd love to add this to the next release, Copilot 4.3.

**Comment by fdedden** [comment]
@nadiaYvette made a very useful contribution during Zurihac this year, but we never came around to incorporate the changes. Basically it fixes most of the compilation warnings we have. That should make overall development of Copilot nicer, the output won't be cluttered so much with warnings.

The changes suggested by @nadiaYvette are a good start, but we definitely want to split it into PR's for the separate Copilot packages.
